addons/delivery/models/product_template.py

```diff
@@ -11,3 +12,13 @@ class ProductTemplate(models.Model):
         string="HS Code",
         help="Standardized code for international shipping and goods declaration. At the moment, only used for the FedEx shipping provider.",
     )
+
+    @api.constrains('sale_ok', 'purchase_ok')
+    def _check_is_delivery(self):
+        for record in self:
+            if record.sale_ok or record.purchase_ok:
+                carriers = self.env['delivery.carrier'].sudo().\
+                    search([('product_id', 'in', record.product_variant_ids.ids)])
```

Comment:
I would search all delivery products before the loop to make only one db request